dot/state/roottottrie.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package state
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ChainSafe/chaindb"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/trie"
+)
+
+type rootToTrieMap struct {
+	rootToTrie map[common.Hash]*trie.Trie
+	sync.RWMutex
+}
+
+func newRootToTrieMap() *rootToTrieMap {
+	return &rootToTrieMap{
+		rootToTrie: make(map[common.Hash]*trie.Trie),
+	}
+}
+
+func (r *rootToTrieMap) get(root common.Hash) (t *trie.Trie, has bool) {
+	r.RLock()
+	defer r.RUnlock()
+	t, has = r.rootToTrie[root]
+	return t, has
+}
+
+func (r *rootToTrieMap) set(root common.Hash, trie *trie.Trie) {
+	r.Lock()
+	defer r.Unlock()
+	r.rootToTrie[root] = trie
+}
+
+func (r *rootToTrieMap) setIfUnset(root common.Hash, trie *trie.Trie) {
+	r.Lock()
+	defer r.Unlock()
+
+	_, exists := r.rootToTrie[root]
+	if exists {
+		return
+	}
+
+	r.rootToTrie[root] = trie
+}
+
+func (r *rootToTrieMap) delete(root common.Hash) {
+	r.Lock()
+	defer r.Unlock()
+	delete(r.rootToTrie, root)
+}
+
+func (r *rootToTrieMap) getFromMemOrDB(root common.Hash, db chaindb.Database) (
+	t *trie.Trie, err error) {
+	r.Lock()
+	defer r.Unlock()
+
+	var calculatedRoot common.Hash
+	t, has := r.rootToTrie[root]
+	if has {
+		return t, nil
+	}
+
+	t = trie.NewEmptyTrie()
+	err = t.Load(db, root)
+	if err != nil {
+		return nil, fmt.Errorf("cannot load root %s from database: %w", root, err)
+	}
+
+	calculatedRootHash := t.MustHash()
+	if !calculatedRootHash.Equal(root) {
+		panic(fmt.Sprintf("trie does not have expected root, expected %s but got %s",
+			root, calculatedRootHash))
+	}
+
+	r.rootToTrie[calculatedRoot] = t
```

Comment:
```suggestion
	r.set(calculatedRoot, t)
```

Reply:
the whole operation has to be thread safe, this is not possible 😉